common/filesystem.cpp

```diff
@@ -0,0 +1,515 @@
+  }
+  Dir result(result_fd);
+
+  // If we actually created the directory, we also need to verify that the
+  // opened file descriptor continues to have the same permissions and the
+  // correct owner and group as we couldn't do the creation atomically with the
+  // open.
+  if (created) {
+    auto stat_result = result.Stat();
+    if (!stat_result.ok()) {
+      // Manually propagate this error so we can attach it back to the opened
+      // path and relative directory.
+      return PathError(stat_result.error().number(),
+                       "DirRef::Stat after opening '{0}' relative to '{1}'",
+                       std::move(path), dfd_);
+    }
+
+    // Check that the permissions are a subset of the requested ones. They may
+    // have been masked down by `umask`, but if there are *new* permissions,
+    // that would be a security issue. We first need to extract the permission
+    // bits from the mode, as other bits are separately controlled.
+    if ((stat_result->permissions() & creation_mode) !=
+        stat_result->permissions()) {
+      // Model this `EPERM`.
+      return PathError(EPERM,
+                       "Setting permissions when creating '{0}' relative to "
+                       "'{1}' during DirRef::OpenDir",
+                       std::move(path), dfd_);
+    }
+    // Also check that the UID is the current effective UID. We don't currently
+    // verify the GID because it could come from the parent directory, so
+    // callers that need to should instead validate this themselves.
+    if (stat_result->uid() != geteuid()) {
+      // Model this as `EPERM`, which is a bit awkward, but should be fine.
+      return PathError(EPERM,
+                       "Setting UID when creating '{0}' relative to "
+                       "'{1}' during DirRef::OpenDir",
+                       std::move(path), dfd_);
+    }
+  }
+
+  return result;
+}
+
+auto MakeTmpDir() -> ErrorOr<RemovingDir, Error> {
+  std::filesystem::path tmpdir_path = "/tmp";
+  for (const char* tmpdir_env_name : {"TEST_TMPDIR", "TMPDIR"}) {
+    const char* tmpdir_env_cstr = getenv(tmpdir_env_name);
+    if (tmpdir_env_cstr == nullptr) {
+      continue;
+    }
+    std::filesystem::path tmpdir_env = std::string(tmpdir_env_cstr);
+    if (!tmpdir_env.is_absolute()) {
+      continue;
+    }
+    tmpdir_path = std::move(tmpdir_env);
+    break;
+  }
+
+  std::filesystem::path target = BuildData::BuildTarget.str();
+  std::string dir_name = target.filename().native();
+  dir_name += ".XXXXXX";
+
+  tmpdir_path /= dir_name;
+
+  std::string tmpdir_path_buffer = tmpdir_path.native();
+  char* result = mkdtemp(tmpdir_path_buffer.data());
+  if (result == nullptr) {
+    RawStringOstream os;
+    os << llvm::formatv("Calling mkdtemp on '{0}' failed: ",
+                        tmpdir_path.native());
+    PrintErrorNumber(os, errno);
+    return Error(os.TakeStr());
+  }
+  CARBON_CHECK(result == tmpdir_path_buffer.data());
```

Comment:
```suggestion
  CARBON_CHECK(result == tmpdir_path_buffer.data(), "`mkdtemp` used a modified path");
```

IIUC the intent here?

Reply:
I think slightly different, but will add this and tweak.